src/Middleware/RequestThrottling/perf/Microbenchmarks/QueueFullOverhead.cs

```diff
@@ -0,0 +1,79 @@
+using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using BenchmarkDotNet.Attributes;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Logging.Abstractions;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.RequestThrottling.Microbenchmarks
+{
+    public class QueueFullOverhead
+    {
+        private const int _numRequests = 2000;
+        private int _requestCount = 0;
+        private TaskCompletionSource<bool> tcs;
+
+        private RequestThrottlingMiddleware _middleware;
+
+        [Params(1,8,64)]
+        public int MaxConcurrentRequests;
+
+        [GlobalSetup]
+        public void GlobalSetup()
+        {
+            var options = new RequestThrottlingOptions
+            {
+                MaxConcurrentRequests = MaxConcurrentRequests,
+                RequestQueueLimit = _numRequests
+            };
+
+            _middleware = new RequestThrottlingMiddleware(
+                    next: (RequestDelegate)_incrementAndCheck,
+                    loggerFactory: NullLoggerFactory.Instance,
+                    options: Options.Create(options)
+                );
+        }
+
+        [IterationSetup]
+        public void Setup()
+        {
+            _requestCount = 0;
+            tcs = new TaskCompletionSource<bool>();
```

Comment:
Nit: we usually use `object` as `T` when it doesn't matter

If this isn't being awaited, we can just use a ManualResetEventSlim, then you don't need to allocate in Setup().